src/Inputs/scene.inputManager.ts

```diff
@@ -804,64 +806,81 @@ export class InputManager {
             }
         };
 
-        let attachedFunction = () => {
-            if (!elementToAttachTo || this._keyboardIsAttached) {
-                return;
-            }
-            elementToAttachTo.addEventListener("keydown", this._onKeyDown, false);
-            elementToAttachTo.addEventListener("keyup", this._onKeyUp, false);
-            this._keyboardIsAttached = true;
-        };
+        this._deviceSourceManager.onDeviceConnectedObservable.add((device) => {
+            // Keyboard Events
+            if (device.deviceType == DeviceType.Keyboard) {
+                this._deviceSourceManager.getDeviceSource(device.deviceType, device.deviceSlot)!.onInputChangedObservable.add((deviceSource) => {

```

Comment:
why do we need this?
this._deviceSourceManager.getDeviceSource(device.deviceType, device.deviceSlot)

from an API standpoint I would expect to be able to do device.onInputChangedObservable...

Reply:
That's an old change from before my last pushed PR.  Good catch!  I'll change it

Reply:
With the current contract, the problem with using `device` directly is that even within that if statement, TypeScript has not inferred that `device` is specifically `DeviceSource<DeviceType.Keyboard>` and so methods on it like `getInput` are not type safe. On the other hand, if you call `deviceSourceManager.getDeviceSource` the returned value's exact type *is* inferred, and it is type safe. I think the problem with the current contract is that `DeviceSourceManager.onDeviceConnectedObservable` is of type `Observable<DeviceSource<DeviceType>>`. We could introduce a new "distributive conditional type" that expands to a union type of all the possible specific `DeviceSource<T>` types, and then TypeScript would properly infer the type of `device` within the `if` check. It would look like this:

```typescript
type DistributedDeviceSource<T> = T extends DeviceType ? DeviceSource<T> : never;
export type KnownDeviceSource = DistributedDeviceSource<DeviceType>;
...
public readonly onDeviceConnectedObservable = new Observable<KnownDeviceSource>((observer) => { ... }
```

In this case, `KnownDeviceSource` resolves to `DeviceSource<DeviceType.Generic> | DeviceSource<DeviceType.Keyboard> | DeviceSource<DeviceType.Mouse> | DeviceSource<DeviceType.Touch> | DeviceSource<DeviceType.DualShock> | DeviceSource<DeviceType.Xbox> | DeviceSource<DeviceType.Switch>`. This would be fully type safe, but would introduce what appears to be a new type (`KnownDeviceSource`). It would also be backward compatible with what we currently have.

A simpler alternative would have been to have the event payload of `onDeviceConnectedObservable` only expose the `deviceType` and `deviceSlot` and force the user to retrieve the device from `DeviceSourceManager.getDeviceSource`, but we can't change that now without breaking backward compatibility.